server-base/src/main/java/org/apache/kylin/rest/service/QueryService.java

```diff
@@ -605,6 +618,29 @@ public SQLResponse searchQueryInCache(SQLRequest sqlRequest) {
         if (response == null) {
             return null;
         }
+
+        // Check whether duplicate query exists
+        while (response.isRunning()) {
+            // Wait at most one minute
+            if (System.currentTimeMillis() - response.getSignature() >= 60000L) {
+                queryCache.evict(sqlRequest.getCacheKey());
+                return null;
+            }
+            logger.info("Duplicated SQL request is running, waiting...");
+            try {
+                Thread.sleep(100);
```

Comment:
DONOT ignore InterruptedException, refer to https://www.ibm.com/developerworks/cn/java/j-jtp05236.html

Reply:
Thanks Dong. I'll throw a RuntimeException for it.